staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,417 @@
+		}
+		if event != nil {
+			t.Errorf("unexpected event, expected a nil event, got %v", *event)
+		}
+	}
+}
+
+func TestCacheIntervalNextFromWatchCache(t *testing.T) {
+	// Have the capacity such that it facilitates
+	// filling the interval buffer more than once
+	// completely and then some more - 10 here is
+	// arbitrary.
+	const capacity = 2*bufferSize + 10
+	wc := newTestWatchCache(capacity, &cache.Indexers{})
+	for i := 0; i < capacity; i++ {
+		wc.Add(
+			makeTestPod(
+				fmt.Sprintf("pod%d", i+1),
+				uint64(i+1),
+			),
+		)
+	}
+
+	cases := []struct {
+		startIndexWatchCache int
+		endIndexWatchCache   int
+		// offset signifies the offset in the watchCache
+		// starting from which events will be served. The
+		// startIndex of the watchCacheInterval is computed
+		// as startIndexWatchCache + offset which means when
+		// the offset is not specified or is 0, the interval
+		// spans the entire watchCache.
+		offset int
+	}{
+		// watchCache empty, startIndexWatchCache = endIndexWatchCache.
+		{startIndexWatchCache: 0, endIndexWatchCache: 0},
+		// watchCache not empty, starts at 0 and contains elements to fill
+		// the interval buffer completely once and then some more with no offset.
+		{startIndexWatchCache: 0, endIndexWatchCache: bufferSize + 10},
+		// same as above but with offset.
+		{startIndexWatchCache: 0, endIndexWatchCache: bufferSize + 10, offset: 5},
+		// watchCache not empty, starts at 0 and contains elements to fill
+		// the interval buffer completely twice and then some more.
+		{startIndexWatchCache: 0, endIndexWatchCache: capacity},
+		// same as above but with offset.
+		{startIndexWatchCache: 0, endIndexWatchCache: capacity, offset: 5},
+		// watchCache non-empty, starts at non-zero index and contains elements
+		// to fill the interval buffer completely once and then some more.
+		{startIndexWatchCache: 10, endIndexWatchCache: bufferSize + 10},
+		// same as above but with offset.
+		{startIndexWatchCache: 10, endIndexWatchCache: bufferSize + 10, offset: 5},
+		// watchCache not empty, starts at non-zero index and contains elements
+		// to fill the interval buffer completely twice and then some more.
+		{startIndexWatchCache: 10, endIndexWatchCache: capacity},
+		// same as above but with offset.
+		{startIndexWatchCache: 10, endIndexWatchCache: capacity, offset: 5},
+	}
+	indexerFunc := func(i int) *watchCacheEvent {
+		return wc.cache[i%wc.capacity]
+	}
+
+	for _, c := range cases {
+		wci := newCacheInterval(
+			c.startIndexWatchCache+c.offset,
+			c.endIndexWatchCache,
+			indexerFunc,
+			wc.isIndexValid,
+			&wc.RWMutex,
+		)
+
+		// Simulate and test interval invalidation iff
+		// the watchCache itself is not empty.
+		if c.startIndexWatchCache < c.endIndexWatchCache {
+			// With an empty interval buffer, simulate the interval being
+			// invalidated before starting to retrieve events.
+			originalCacheStartIndex := wc.startIndex
+			wc.startIndex = wci.startIndex + 1
```

Comment:
We should test that not just at the beginning, but also (which would probably happen more frequently) when the invalidations happens somewhere in the middle of processing.

I suggest adding a test param for "invalidateAfterNEvents" or sth like that.